src/memory_target_ops.h

```diff
@@ -0,0 +1,48 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#ifndef UMF_MEMORY_TARGET_OPS_H
+#define UMF_MEMORY_TARGET_OPS_H 1
+
+#include <stdbool.h>
+
+#include <umf/base.h>
+#include <umf/memspace.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct umf_memory_target_t *umf_memory_target_handle_t;
+
+struct umf_memory_target_ops_t {
+    /// Version of the ops structure.
+    /// Should be initialized using UMF_VERSION_CURRENT
+    uint32_t version;
+
+    enum umf_result_t (*initialize)(void *params, void **mem_target);
+    void (*finalize)(void *mem_target);
+
+    enum umf_result_t (*pool_create_from_memspace)(
+        umf_memspace_handle_t memspace, void **mem_targets, size_t num_targets,
+        umf_memspace_policy_handle_t policy, umf_memory_pool_handle_t *pool);
+
+    enum umf_result_t (*memory_provider_create_from_memspace)(
```

Comment:
Interesting design question: Who (which entity) should be responsible for choosing a pool manager?
If a user calls `umfMemoryProviderCreateFromMemspace` then the user will explicitly specify the pool manager when the `umfPoolCreate` function is called.
I am mainly interested in the case when a user calls 'umfPoolCreateFromMemspace'. There possible candidates:

- Memspace implementation chooses the heap manager
- UMF implementation chooses the heap manager based on the memspace or memory provider instance.

Another consideration to keep in mind is that some heap managers are provided as static libraries (not part of libumf).

Let's think about that and discuss it in the next sync.

Reply:
We have 2 entities that impact which pool/provider should be used: memspace (i.e collection of memory_targets) and policy.

In C++, we could probably make it a 'cusomization point' and let user/developer just define this function for each combination of policy/memory targets he's interested in. In C, I think we have to put this logic either in memory_target_ops or in policy somehow. I've chosen to put it in memory_target_ops since it's easier for now (we already have the ops abstraction). However, I actually belive that policy would be a better place (since it would allow us to mix tagets).

We could implement policy using `policy_ops` structure (similarly as memory_target, pool, provider) and have those `pool_from_memspace` and `provider_from_memspace` be part of that. We can then have a default policy that would be a fallback and also allow users or memory_target developers to implement their own policies that would have some custom logic for choosing the right pool/provider. This should work pretty well with tiering for example: we would have a 'tiered' policy and the core logic would be just in `pool_from_memspace` in that policy.

Reply:
@vinser52 @bratpiorka I made the pool_create_from_memspace return NOT_SUPPORTED for now. This function is not strictly needed for now.